pkg/v1/providers/ytt/lib/helpers.star

```diff
@@ -79,15 +79,13 @@ def tkg_image_repo():
 end
 
 def get_image_repo_for_component(image):
-  # if custom image repo is specified use that repo
-  if data.values.TKG_CUSTOM_IMAGE_REPOSITORY:
-    return data.values.TKG_CUSTOM_IMAGE_REPOSITORY
-  end
-
   # if imageRepo is specified for the component with image.imageRepository
   # than use image.imageRepository else use default imageRepo from BoM file
   if hasattr(image, 'imageRepository'):
     return image.imageRepository
+  # if custom image repo is specified use that repo
+  elif data.values.TKG_CUSTOM_IMAGE_REPOSITORY:
+    return data.values.TKG_CUSTOM_IMAGE_REPOSITORY
```

Comment:
@vuil yeah seems like that would work. I am sad that it does feel like the overriding is getting increasing complex i.e the mental model I'm building in my head isn't easily described. I definitely agree with @tylerschultz that the mental model of the level of specificity defining the order is easier to reason about and preferable, but I also understand that current usage dictates what can be changed. I probably need to stew on this a bit more, but I'd definitely be interested in the answer to @tylerschultz's question.

> Are we trying to provide custom images for Antrea? That is currently driven by carvel core package repository and I am not sure if this image specific override will achieve what you are trying to do.

This is just overriding images in general (but testing CNI changes is definitely one of them), since we do end up touching a different images depending on what we are working on. Also I think it would still be useful for things like Antrea where it initially deploys using the `providers` and then pivots to packages (although I do think it would still be great if we can get to the point of only deploying this with packages). Often times the initial deploy needs to work and it is useful to switch the image then.

Reply:
@anujc25 regarding this concern:

> if the production BoM file specifies `image.imageRepository` 

I took a look at the latest production BoM's and from what I can tell, they do _not_ specify `image.imageRepository` anywhere. Have we done that in the past? Do we expect to do that in the future? Why?

If that was just a "what if" scenario that we've never seen a real need to support, then I think we should consider merging this as-is as. 

---

On the other hand, perhaps it's worth re-examining this assumption from the original post:

> My current workflow requires me to override the global image repository using `TKG_CUSTOM_IMAGE_REPOSITORY`, which I imagine is pretty common if you don't want to use the standard image repository.

Why do we need to use that environment variable in our development workflows? Do local builds not default to the registry we want? If not, and we have to edit the BoM anyway to override a specific image's registry, couldn't we also edit the BoM's global registry to be what we want at the same time? Will look into that as well...

Reply:
I am concerned if a dx enhancement leads to something that breaks product or renders something really hard to workaround so we need to be careful about it.

> I took a look at the latest production BoM's and from what I can tell, they do _not_ specify `image.imageRepository` anywhere. Have we done that in the past? Do we expect to do that in the future? 

It's a good point. Our boms indeed do not have this situation, since every component we referenced in the BOM have been built and pushed to a location we have access to. 

> If that was just a "what if" scenario that we've never seen a real need to support, then I think we should consider merging this as-is as. 

The problematic scenario that 'could' happen is if both of these situations occur.
1. There are some customization done to a BOM (a component is introduced, and since the introducer does not have access to default registry, component is pushed to another registry, which means image.respository has to be set for it in the BOM)
2. The user of the BOM is someone who does not want to pull from any public repo, so need an airgap setup and leverage TKG_CUSTOM_IMAGE_REPOSITORY to point to a private repo.

The workaround I suppose is for the introducer to modify the bom 
1. at step 1 so that all images are pulled from a single repo, eliminating the need to override the repo for a single component or
2. the user to modify the BOM at step 2 to delete the repository referenced for that one specific component so the TKG_CUSTOM_IMAGE_REPOSITORY override will cover it.